tests/Unit/Framework/Pypp.hpp

```diff
@@ -21,262 +22,311 @@
+          container_dv[storage_index], grid_point_index);
+    }
+    return container_double;
+  }
+
+  static inline void pack(const gsl::not_null<Tensor<T, Ts...>*> packed_t,
+                          const unpacked_type& t,
+                          const size_t grid_point_index) {
+    for (size_t storage_index = 0; storage_index < t.size(); ++storage_index) {
+      ContainerPackAndUnpack<T>::pack(
+          make_not_null(&(*packed_t)[storage_index]), t[storage_index],
+          grid_point_index);
+    }
+  }
+
+  static inline Tensor<T, Ts...> construct_zero_result(const size_t num_pts) {
+    return Tensor<T, Ts...>{
+        ContainerPackAndUnpack<T>::construct_zero_result(num_pts)};
+  }
+
+  static inline size_t get_size(const Tensor<T, Ts...>& t) noexcept {
+    return ContainerPackAndUnpack<T>::get_size(t[0]);
+  }
 };
 
-template <typename T>
-using convert_to_container_of_doubles_t =
-    typename convert_to_container_of_doubles<T>::type;
+template <typename T, size_t Size>
+struct ContainerPackAndUnpack<std::array<T, Size>, std::nullptr_t> {
+  using unpacked_type =
+      std::array<typename ContainerPackAndUnpack<T>::unpacked_type, Size>;
+  using packed_type =
+      std::array<typename ContainerPackAndUnpack<T>::packed_type, Size>;
+  using underlying_packed_type =
+      typename ContainerPackAndUnpack<T>::packed_type;
 
-template <typename T>
-struct SliceContainerImpl {
-  static auto apply(const T& container_dv, const size_t slice_idx) noexcept {
-    convert_to_container_of_doubles_t<std::decay_t<decltype(container_dv)>>
-        container_double{};
-    ASSERT(slice_idx < container_dv.begin()->size(),
-           "Trying to slice DataVector of size " << container_dv.begin()->size()
-                                                 << "with slice_idx "
-                                                 << slice_idx);
-    for (decltype(auto) double_and_datavector_components :
-         boost::combine(container_double, container_dv)) {
-      boost::get<0>(double_and_datavector_components) =
-          boost::get<1>(double_and_datavector_components)[slice_idx];
+  static inline unpacked_type unpack(const std::array<T, Size>& container_dv,
+                                     const size_t grid_point_index) noexcept {
+    unpacked_type container_double{};
+    for (size_t i = 0; i < Size; ++i) {
+      gsl::at(container_double, i) = ContainerPackAndUnpack<T>::unpack(
+          gsl::at(container_dv, i), grid_point_index);
     }
     return container_double;
   }
+
+  static inline void pack(const gsl::not_null<std::array<T, Size>*> packed_t,
+                          const unpacked_type& t,
+                          const size_t grid_point_index) {
+    for (size_t i = 0; i < t.size(); ++i) {
+      ContainerPackAndUnpack<T>::pack(make_not_null(&gsl::at(*packed_t, i)),
+                                      gsl::at(t, i), grid_point_index);
+    }
+  }
+
+  static inline std::array<T, Size> construct_zero_result(
```

Comment:
Picking an arbitrary one: Could `make_with_value` be used instead of `construct_zero_result` and `get_size`?

Reply:
My worry would be that some classes like Equations of State don't have a `make_with_value`, but they also wouldn't really be returned from python functions. But even that's not fully supported, so yea, I think switching to `make_with_value` works.

Reply:
`get_size` is used for looping over the grid points, though. I'm not sure how to get around that. `number_of_points` is private (in a detail namespace) and it looks like not required to be implemented for all types(?)